### PR TITLE
Fix PubSub timeout propagation to prevent indefinite hangs on socket read operations

### PR DESCRIPTION
**Pull request by petyaslavova**
# 🛠 Fix: Prevent Indefinite Hang in `PubSub.get_message()` with Timeouts

## 🚨 Problem

This PR fixes a **critical hang issue** in `PubSub.get_message()` when using timeouts.

Previously:

- `can_read()` correctly respected the provided `timeout`.
- However, the subsequent `read_response()` call did **not** receive the timeout.
- As a result, it fell back to the default socket timeout (often **infinite**).

### ❗ Consequence

This created a scenario where:

1. `can_read()` reported that data was available.
2. `read_response()` attempted to read from the socket.
3. The socket read blocked indefinitely (e.g., due to partial data or connection state changes).

This resulted in **unexpected and indefinite hangs**.

---

## ✅ Solution

This PR propagates an optional `timeout` parameter through the entire call chain:
get_message()
→ read_response()
→ parser layer
→ connection layer
→ socket-level recv()

### Coverage Includes

- **All parser implementations**
  - Hiredis
  - RESP2
  - RESP3

- **All connection types**
  - Standard connections
  - Sentinel-managed connections
  - Cache-proxy connections

- **Both execution models**
  - Synchronous
  - Asynchronous

The timeout is now consistently respected down to the lowest-level socket `recv()` call.

---

## 🧪 Test Coverage

This PR adds **comprehensive test coverage (887 lines)** validating timeout behavior across:

- Cluster sharded Pub/Sub
- Pattern subscriptions
- Standard Pub/Sub flows
- Edge cases:
  - `timeout=0`
  - `timeout=None`

All relevant sync and async paths are covered.

---

## 🎯 Result

`PubSub.get_message()` now behaves correctly and predictably when timeouts are used, eliminating the risk of indefinite blocking under partial-read or connection edge cases.

---

> [!NOTE]
> **Medium Risk**
> Touches core response parsing and socket read paths (RESP2/RESP3/Hiredis and `Connection.read_response`), so mistakes could cause new timeouts or regressions in message parsing under load. Added/updated tests reduce risk, but behavior changes affect all PubSub and push-response reads.
> 
> **Overview**
> Fixes Pub/Sub timeouts so `get_message(timeout=...)` can’t hang indefinitely by **threading an explicit `timeout` value through `PubSub.parse_response()` → `Connection.read_response()` → parser implementations → `SocketBuffer` socket reads**.
> 
> Adds a `timeout` parameter (defaulting to `SENTINEL` to preserve existing socket timeout behavior) across RESP2/RESP3/Hiredis parsers, `SocketBuffer.read/readline`, `SentinelManagedConnection`, and cache proxy connections, and propagates the timeout in cluster sharded Pub/Sub message polling.
> 
> Expands test coverage with new sync/async PubSub and cluster sharded PubSub timeout-propagation tests, plus unit tests verifying timeout passthrough for proxy and sentinel connections.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 32c19a079d91ff87e16688cd19975d3ec4e538b4. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: redis/_parsers/hiredis.py [hunk 1]
```diff
@@ -132,7 +132,12 @@ def read_from_socket(self, timeout=SENTINEL, raise_on_timeout=True):
             if custom_timeout:
                 sock.settimeout(self._socket_timeout)
 
-    def read_response(self, disable_decoding=False, push_request=False):
+    def read_response(
+        self,
+        disable_decoding=False,
+        push_request=False,
+        timeout: Union[float, object] = SENTINEL,
```
Comment (vladvildanov) [inline redis/_parsers/hiredis.py:139]: Object as timeout?

Comment (petyaslavova) [reply redis/_parsers/hiredis.py:139]: We need to be able to distinguish between the cases when arg is just  not provided, and None is intentionally set.
